### PR TITLE
Add the ability to make HTTP requests

### PR DESCRIPTION
**Pull request by TylerKnvpp**
### Background
This change was implemented to enhance Auto-GPT's capabilities by enabling it to interact with external APIs and web services through HTTP requests. The addition of this feature allows the AI to access a broader range of information and resources, opening up new possibilities for task completion and automation.

### Changes
In this pull request, a focused and specific change has been made to enable Auto-GPT to perform HTTP requests. The following modifications were implemented:

1. **Created the make_http_method:** A new function called `make_http_method` was developed. This function is responsible for handling and executing the actual HTTP requests based on the given method type, authentication token, input, and URL.
2. **Added the command:** The command that triggers the `make_http_method` function was integrated into Auto-GPT's existing command structure. This allows users to invoke the new functionality by simply providing the required parameters.

These modifications were specifically focused on enhancing Auto-GPT's capabilities by adding the ability to make HTTP requests. No unrelated or "extra" changes were included in this pull request, ensuring a clear and concise implementation.

### Documentation
The changes made in the provided method are documented through a combination of in-code comments and clear function naming conventions. This ensures that the code is easy to understand and maintain for other contributors or users. Here's a breakdown of the documentation:

1. **Import statements:** Necessary libraries and modules are imported at the beginning of the code, making it clear which dependencies are required for the method to function correctly.
2. **Type alias for JSON:** A type alias named JSONType is created to represent various JSON data types. This makes it easier to understand the expected data types when handling JSON data within the function.
3. **Function definition:** The `make_http_request` function is defined with a clear and descriptive name, along with typed arguments (url, method, and data) that indicate the expected input parameters. This makes it easy to understand the function's purpose and usage.
4. **In-code comments:** Throughout the function, concise comments are placed to explain each step of the process, such as loading the API key, setting up headers, and sending the request. These comments guide the reader through the code, making it easier to understand the logic and flow.
5. **Error handling:** The try and except blocks are used to handle any errors that might occur during the request process. In the event of an error, the code prints a message and a traceback to provide context for debugging.

### Test Plan
1. **Setup an API route:** I created an API route within another application that listens for incoming requests at `/api/auto-gpt`. This serves as a test endpoint to validate the functionality of the `make_http_request` method.
2. **Create test cases:** I prepared several test cases, with the most relevant being a POST request to http://localhost:3000/api/auto-gpt with a request body containing a message saying "Hello, World".
3. **Run the test:** To test the functionality, I gave the agent the following tasks
  - Prompt the user for a url, what they want to send, and how to authorize the request
  - format what the user whats to send in JSON
  - use the `make_http_request` to make an HTTP request to the desired URL
4. **Verify the response:** After sending the request, I checked the response received from the `/api/auto-gpt` endpoint. I verified that the response had the expected status code (e.g., 200 OK or 201 Created), and the response body contained the expected data or an acknowledgment of the successful POST request by logging it to the terminal.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Review by Pwuts — CHANGES_REQUESTED** [review]
1. Please change the base branch to `master`
2. Most of this functionality is already in `web_request.py`. Try mending that to be available as a command.

File: autogpt/commands/web_requests.py [hunk 1]
```diff
@@ -4,20 +4,54 @@
...
+
+@command(
+    "make_http_request",
+    "Make an HTTP request",
+    '"url": "<url>", "method": "<method>", "auth_token": "<auth_token>", "data": "<data>"',
+)
+def make_http_request(url: str, method: str, auth_token: str, data: JSONType):
+    valid_url = is_valid_url(url)
+
+    if not valid_url:
```
Comment (Pwuts) [inline autogpt/commands/web_requests.py]:
```suggestion
    if not is_valid_url(url):
```

File: autogpt/commands/web_requests.py [hunk 2]
```diff
@@ -4,20 +4,54 @@
...
 
 session = requests.Session()
 session.headers.update({"User-Agent": CFG.user_agent})
 
+JSONType = Union[Dict[str, Any], List[Any], str, int, float, bool, None]
+
+@command(
+    "make_http_request",
+    "Make an HTTP request",
+    '"url": "<url>", "method": "<method>", "auth_token": "<auth_token>", "data": "<data>"',
```
Comment (Pwuts) [inline autogpt/commands/web_requests.py]: I'm not sure how, but it needs to be clear from this spec that all parameters except `url` are optional.

File: autogpt/commands/web_requests.py [hunk 3]
```diff
@@ -4,20 +4,54 @@
...
+)
+def make_http_request(url: str, method: str, auth_token: str, data: JSONType):
+    valid_url = is_valid_url(url)
+
+    if not valid_url:
+        return "Error: Invalid URL"
+
+    sanitized_url: str = sanitize_url(url)
+
+    print("Starting HTTP request...")
```
Comment (Pwuts) [inline autogpt/commands/web_requests.py]: Please use logger with info level for this. Logging in the application is currently a mess and we have to start improving somewhere. :)

File: autogpt/commands/web_requests.py [hunk 4]
```diff
@@ -4,20 +4,54 @@
...
+    # Set up headers with API key
+    headers = {
+        "Authorization": f"Bearer {auth_token}",
+        "Content-Type": "application/json",
+    }
```
Comment (Pwuts) [inline autogpt/commands/web_requests.py]: `Authorization` should only be set if a token is passed; `Content-Type` should only be set if the request is `POST`/`PUT` or alike.

**Review by Pwuts — CHANGES_REQUESTED** [review]
I have a few comments/requests, but overall it looks pretty good!

In addition to the comments below: can you add a unit test for `make_http_request`?

**Comment by TylerKnvpp** [comment]
@Pwuts Feedback addressed! To run the unit tests, please run: `python3 -m unittest tests/unit/test_make_http_request.py`

**Comment by Pwuts** [comment]
Have you tested whether the LLM understands the command and the optional parameters?

**Comment by TylerKnvpp** [comment]
> Have you tested whether the LLM understands the command and the optional parameters?

@Pwuts I have run a series of tests, evaluating various methods and conditionally incorporating authentication tokens within the prompts. The following examples illustrate that the authentication token and request body are acknowledged as optional.

_GET with auth token_
Screenshot 2023-04-24 at 8 38 23 AM

_GET without auth token_
Screenshot 2023-04-24 at 9 06 27 AM

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Comment by ntindle** [comment]
This was recently added as a plug-in in the plugins repo! Would you mind merging your contributions into that as it's rather basic right now?

**Pull request closed without merging**